File: largestDivisorSmallerThanThreshold.js
const print = console.log;

function divideAndAdd(nums, divisor) {
  let sum = 0;
  for(let i = 0; i < nums.length; i++) {
      sum += Math.ceil(nums[i] / divisor);
  }
  return sum;
}

function smallestDivisor(nums, threshold) {
  let s = 1;
  let e = Math.max(...nums);

  while(s < e) {
    let m = s + Math.floor((e - s) / 2);
    if(divideAndAdd(nums, m) > threshold) {
      s = m + 1;
    } else {
      e = m;
    }
  }
  return s;
}

// const arr = [44,22,33,11,1];
const arr = [10101, 12121, 21212];
print(smallestDivisor(arr, 1000000));

File: largestDivisorSmallerThanThreshold.cpp
#include <bits/stdc++.h>
using namespace std;

int divideAndAdd(vector<int> &nums, double divisor)
{
  long long int sum = 0;
  for (size_t i = 0; i < nums.size(); i++)
  {
    sum += ceil(nums[i] / divisor);
  }
  return sum;
}

int smallestDivisor(vector<int> &nums, int threshold)
{
  long long int s = 1;
  long long int e = *max_element(nums.begin(), nums.end());

  while (s < e)
  {
    int m = s + (e - s) / 2;
    if (divideAndAdd(nums, m) > threshold)
    {
      s = m + 1;
    }
    else
    {
      e = m;
    }
  }
  return s;
}

int main()
{
  vector<int> vec = {21212, 10101, 12121};
  long long int threshold = 1000000;
  cout << smallestDivisor(vec, threshold) << endl;
  return 0;
}